src/metabase/api/activity.clj

```diff
@@ -107,31 +125,122 @@
+  viewed then hydrate the basic details of the model.
+
+  Viewing a Dashboard will add entries to the view log for all cards on that dashboard so all card views are instead derived
+  from the query_execution table. The query context is always a `:question`. The results are normalized and concatenated to the
+  query results for dashboard and table views."
+  [views-limit card-runs-limit]
+  (let [dashboard-and-table-views (db/select [ViewLog :user_id :model :model_id
+                                              [:%count.* :cnt] [:%max.timestamp :max_ts]]
+                                    {:group-by [:user_id :model :model_id]
+                                     :where    [:and
+                                                [:= :user_id *current-user-id*]
+                                                [:in :model #{"dashboard" "table"}]]
+                                     :order-by [[:max_ts :desc]]
+                                     :limit    views-limit})
+        card-runs (->> (db/select [QueryExecution [:executor_id :user_id] [:card_id :model_id]
+                                   [:%count.* :cnt] [:%max.started_at :max_ts]]
+                         {:group-by [:executor_id :card_id :context]
+                          :where [:and
+                                  [:= :executor_id *current-user-id*]
+                                  [:= :context (hx/literal :question)]]
+                          :order-by [[:max_ts :desc]]
+                          :limit card-runs-limit})
+                       (map #(dissoc % :row_count))
+                       (map #(assoc % :model "card")))]
+    (->> (concat card-runs dashboard-and-table-views)
+         (sort-by :max_ts)
+         reverse)))
+
 (defendpoint GET "/recent_views"
-  "Get the list of 5 things the current user has been viewing most recently. Query takes 8 and limits to 5 so that if it
-  finds anything archived, deleted, etc it can hopefully still get 5."
+  "Get the list of 5 things the current user has been viewing most recently."
   []
-  ;; expected output of the query is a single row per unique model viewed by the current user
-  ;; including a `:max_ts` which has the most recent view timestamp of the item and `:cnt` which has total views
-  ;; and we order the results by most recently viewed then hydrate the basic details of the model
-  (let [views (db/select [ViewLog :user_id :model :model_id
-                          [:%count.* :cnt] [:%max.timestamp :max_ts]]
-                         {:group-by [:user_id :model :model_id]
-                          :where    [:and
-                                     [:= :user_id *current-user-id*]
-                                     [:in :model #{"card" "dashboard" "table"}]]
-                          :order-by [[:max_ts :desc]]
-                          :limit    8})
+  (let [views (views-and-runs 8 8)
```

Comment:
normally want to create some defs for these two `8`s so its clear what they are and where to tweak them if necessary.

Reply:
I was considering using keyword args to make it clear, but keep the values at the call sites?

I could also make top level defs, like you say. If I do that, should I make `views-and-runs` a 0-arity fn?

Reply:
yeah i would just `(def last-view-count 8)` or something similar. Verbose is ok. Just make it easy for those constants to stand out if anyone needs to change in the future.